### ISSUE TITLE
Parameters of the form type[typevar] cause issues with type refinement

### ISSUE THREAD

**Issue by Pluviolithic** [microsoft/pyright#9568]
**Description**

When defining a function with the `type` of a `TypeVar` as one of the parameters and a return value of that same `TypeVar`, Pyright fails to accurately recognize the return values as being assignable to the return type.

**Code**

```python
from typing import TypeVar, reveal_type

Test = TypeVar(
    "Test", str, int
)

int_arr = [1, 2, 3]
str_arr = ["a", "b", "c"]

def get_test(idx: int, t: type[Test]) -> Test:

    if isinstance(t, int):
        reveal_type(t)
        return int_arr[idx]
    else:
        reveal_type(t)
        return str_arr[idx]
```

Pyright output:
```
test.py:11:17 - information: Type of "t" is "type[Test@get_test]"
test.py:13:21 - information: Type of "t" is "<subclass of type[str]* and int>* | <subclass of type[int]* and int>*"
test.py:14:16 - error: Type "int" is not assignable to return type "Test@get_test"
Type "int" is not assignable to type "Test@get_test" (reportReturnType)
test.py:16:21 - information: Type of "t" is "type[str]* | type[int]*"
test.py:17:16 - error: Type "str" is not assignable to return type "Test@get_test"
Type "str" is not assignable to type "Test@get_test" (reportReturnType)
2 errors, 0 warnings, 3 informations
```

With MyPy, a simple fix is to simply call `t` on the return value. But that's redundant code and shouldn't be a requirement for properly utilizing a typechecker. Furthermore, that same fix does not work with Pyright:

```py
from typing import TypeVar, reveal_type

Test = TypeVar(
    "Test", str, int
)

int_arr = [1, 2, 3]
str_arr = ["a", "b", "c"]

def get_test(idx: int, t: type[Test]) -> Test:
    reveal_type(t)
    if isinstance(t, int):
        reveal_type(t)
        return t(int_arr[idx])
    else:
        reveal_type(t)
        return t(str_arr[idx])
```

Pyright output:
```
test.py:11:17 - information: Type of "t" is "type[Test@get_test]"
test.py:13:21 - information: Type of "t" is "<subclass of type[str]* and int>* | <subclass of type[int]* and int>*"
test.py:14:16 - error: Object of type "<subclass of type[int]* and int>*" is not callable
Attribute "__call__" is unknown (reportCallIssue)
test.py:14:16 - error: Object of type "<subclass of type[str]* and int>*" is not callable
Attribute "__call__" is unknown (reportCallIssue)
test.py:14:16 - error: Return type is unknown (reportUnknownVariableType)
test.py:16:21 - information: Type of "t" is "type[str]* | type[int]*"
3 errors, 0 warnings, 3 informations
```

**Command-line**

Running via command line:
```
.venv ❯ pyright --version
pyright 1.1.390
```

**Comment by erictraut** [comment]
This code doesn't make sense to me. It is checking to see if `t` is an instance of `int`, but `t` is declared to be an instance of `type`. Pyright is doing its best to make sense out of the code by assuming a class that subclasses from both `type` and `int`, but I don't think such a class can be created in the Python runtime. Perhaps you intended to use `issubclass` rather than `isinstance`?

In any case, I think pyright is working correctly here, so I don't consider this a bug.

**Comment by erictraut** [comment]
I noticed that you filed a similar bug in the mypy project. I recommend that you close that one also.

**Comment by Pluviolithic** [comment]
Appreciate it, thanks.

**Comment by erictraut** [comment]
There are a couple of problems here. First, using `t == int` checks specifically for `int` and not a subclass of `int` whereas the annotation `int` means "an instance of `int` or any subclass thereof". It's generally problematic to use equality to check for types at runtime because this approach won't work for subclasses.

Second, you're using a value-constrained type variable here in an odd way. I recommend against using value-constrained type variables unless you _really_ know what you're doing and there is no other way to accomplish your goal. Their behavior is poorly specified in the Python typing spec, and there are odd behaviors and limitations that you will hit. See [this documentation](https://microsoft.github.io/pyright/#/mypy-comparison?id=value-constrained-type-variables) for more details. 

It's still not entirely clear to me what you're trying to do with this code, but an `@overload` may be a more appropriate approach. 

```python
from typing import overload

int_arr = [1, 2, 3]
str_arr = ["a", "b", "c"]

@overload
def get_test(idx: int, t: type[int]) -> int: ...
@overload
def get_test(idx: int, t: type[str]) -> str: ...
def get_test(idx: int, t: type[int | str]) -> int | str:
    if issubclass(t, int):
        return int_arr[idx]
    else:
        return str_arr[idx]
```

It looks like `t` is acting like a boolean flag to select between `int` and `str`. If that's the case, it may be better to simply use a `bool` parameter.

```python
@overload
def get_test(idx: int, use_int: Literal[True]) -> int: ...
@overload
def get_test(idx: int, use_int: Literal[False]) -> str: ...
def get_test(idx: int, use_int: bool) -> int | str:
    if use_int:
        return int_arr[idx]
    else:
        return str_arr[idx]
```

**Comment by Pluviolithic** [comment]
For future users who come by this issue, I'll try to write up a better summary of what is going on as a whole.

The objective is to have some kind of store or registry which contains a myriad of types. In the getter function, the goal would be to pass in the key and the expected output type. The point of using a value-constrained TypeVar would be to verify that the value you get out is of the type you pass in. Naturally, you can't "prove" that, so some type narrowing has to happen in the `get_value` function for edge case handling or for typecasting.

Whatever the case, it seems that there was a misunderstanding of how to properly do the narrowing. Clearly `issubclass` is the correct choice.

I've included a better code example in [pyright playground](https://pyright-play.net/?code=GYJw9gtgBALgngBwJYDsDmUkQWEMoAqiApgGoCGIANFCMQG7HkA2A%2BvAsQFBcHEDO%2BALyESFEAAouUGVABEfQXJqoYNQSC4BKHpRAAuKMySCA2qqgAfKBoC6UEaYCMNOeWVQATK4BGHgMyuAMZytjwAJsTAUGjEMOwCMBJI4QAehqo0MIYcxKaKMLZaUAC0AHyEiVZQAHJgKMT60rJgAK7CUHrmaWGymNEmqILkKEHEEm1qmCgwWk19fUgD-PytPkHM5CsSU6pzzQsLdIws7CQ7OoeHdDCtIChQkwcLxMz8jc9XtHF3D3UNz1eS0w-CGMBGYwm7XUMBA%2By%2BwJMq3Wm22Uw08K%2BsmOTDYuQunyuN1%2Bj3ahJkr3e8yxMmJ91q9W4NJk5KgXAQIFUEli8RgiQkAAYYXCtEA). Per the reasons given [above](https://github.com/microsoft/pyright/issues/9568#issuecomment-2536516084), this does fail when using `pyright`, but [mypy playground](https://mypy-play.net/?mypy=latest&python=3.12&gist=c249ce2a464bf80f53ff31175c0e135d) says it's correct. That being said, this is potentially better implemented with overloads as suggested, though it does add a fair amount of development time as the number of types goes up.

```python
from typing import TypeVar, reveal_type

Test = TypeVar(
    "Test", int, str
)

arr: list[int | str] = [1, "a", 2, "b", 3, "c"]

def get_test(idx: int, t: type[Test]) -> Test | None:
    out = arr[idx]
    if isinstance(out, int):
        if issubclass(t, int):
            reveal_type(t)
            return out
        else:
            return None
    elif isinstance(out, str):
        if issubclass(t, str):
            reveal_type(t)
            return out
        else:
            return None
            
        
print(get_test(0, str))
```
